### ISSUE TITLE
🐛 Always using light theme on dark background

### ISSUE THREAD

**Issue by EbonJaeger** [dandavison/delta#1663]
After updating to the 0.17.0 release, I can no longer read git diffs due to `delta` using a light theme on my dark terminal background.

My setup consists of Alacritty with Zellij and Fish. I tested on a stock GNOME Terminal as well with un-modified bash, and the problem happens there, as well. My `.gitconfig` options for diffing matches that in the Delta readme.

- [x] Please include the raw text output from git, so that we can reproduce the problem.
      (You can use `git --no-pager` to produce the raw text output.)
    ```diff
    diff --git a/packages/a/adw-gtk3-theme/package.yml b/packages/a/adw-gtk3-theme/package.yml
    index aeb8a25200..06017a5c30 100644
    --- a/packages/a/adw-gtk3-theme/package.yml
    +++ b/packages/a/adw-gtk3-theme/package.yml     @@ -1,8 +1,8 @@
     name       : adw-gtk3-theme
    -version    : '5.2'
    -release    : 12
    +version    : '5.3'
    +release    : 13
     source     :
    -    - https://github.com/lassekongo83/adw-gtk3/archive/refs/tags/v5.2.tar.gz : 1265b9dbc933bf602da1c419e515752bc8d7b9e6f8a09eafd740315d321501d7
    +    - https://github.com/lassekongo83/adw-gtk3/releases/download/v5.3/adw-gtk3v5.3.tar.xz : 2e6e87935bef30936e40d07c7af4fd20754e77917be224f61c4346867196bef0
     license    : LGPL-2.1-only
     homepage   : https://github.com/lassekongo83/adw-gtk3
     component  : desktop.them
    ```
- [x] A screenshot of Delta's output is often helpful also.

![image](https://github.com/dandavison/delta/assets/5157277/17bc897d-9acb-41d6-af9f-3fba69caed05)

**Comment by dandavison** [comment]
Hi @EbonJaeger, sorry about that. Please add `dark = true` or `--dark` to your delta config / invocations to force colors for dark backgrounds.

delta in 0.17.0 should autodetect in the absence of the `light/dark` setting. I wonder if it could be an issue with Zellij cc @bash -- any thoughts?

**Comment by rsmmr** [comment]
I'm seeing the same issue but only when I'm inside a `tmux` shell. Outside `tmux` it works fine.

**Comment by bash** [comment]
Hi @EbonJaeger :wave: Thank you for reporting the bug! 
In your GNOME Terminal setup, are you running Bash inside of Zellij or directly in GNOME Terminal?

I would be surprised if the issue happened in GNOME Terminal directly as it has always correctly reported colors in my experience.

@dandavison From a quick glance and some preliminary tests it does seem like Zellij reports incorrect colors (happens to me both in Alacritty and GNOME Terminal). If will try to dig further next week or so.

**Comment by bash** [comment]
@rsmmr What version of tmux are you running? Color detection should work OK in the latest version from git iirc.

**Comment by EbonJaeger** [comment]
I tried again with GNOME terminal this morning and it is working as expected. Maybe I didn't start a new session or something last night after trying to force the dark theme. Sorry about that. So that just leaves Zellij.

The workaround of setting `dark = true` in the `.gitconfig` works, though! Thanks!

**Comment by bash** [comment]
That's good to hear! I was worried there for a second since GNOME Terminal always worked well in my experience 😆

**Comment by lilyball** [comment]
I've hit this too. It works fine in tmux, but not in zellij. What's more annoying is that the new `--detect-dark-light` flag cannot be set through gitconfig, which means my only option is to force dark on all the time instead of letting it infer from my theme.

**Comment by lilyball** [comment]
Also I will note that Zellij sets a few env vars that can be used to detect that we're running in Zellij.

**Comment by sedlund** [comment]
same, it does not detect.  defaults to light mode, unreadable in dark mode.

delta 0.18.2
bat 0.25.0  ( config: --theme=base16)
tmux 3.5.a

**Comment by dandavison** [comment]
Hi @sedlund, sorry about that. Just to confirm -- you're able to work around by hardcoding `dark` or `light`, right? Just autodetection (under zellij?) is not working?

**Comment by sedlund** [comment]
Hi @dandavison yes, workaround with tmux.  I updated my original reply with the particulars.

**Comment by theCapypara** [comment]
I looked a bit into this, the most plausible reason for detection failing is that the library delta uses to detect the color theme is explicitly [not supporting screen](https://github.com/bash/terminal-colorsaurus/blob/main/crates/terminal-colorsaurus/src/quirks.rs#L27-L55), which also disables detection on tmux, since it is setting the same TERM variable.

**Comment by tautropfli** [comment]
> I looked a bit into this, the most plausible reason for detection failing is that the library delta uses to detect the color theme is explicitly [not supporting screen](https://github.com/bash/terminal-colorsaurus/blob/main/crates/terminal-colorsaurus/src/quirks.rs#L27-L55), which also disables detection on tmux, since it is setting the same TERM variable.

Yeah I just saw that tmux doesn't always have the same `TERM` variable: Depending on some checks done at build time it might use `screen` instead of `tmux` (or `tmux-256color`).

As a workaround, you can always set the `TERM` variable yourself:
```tmux
set -g default-terminal "tmux-256color"
```

**Comment by theCapypara** [comment]
I had issues with some programs when setting that variable to `tmux-256-color`.
I ended up with a (arguably somewhat silly but pretty effective) workaround using a wrapper script that just asks my desktop for the current theme:

```sh
#!/usr/bin/env bash
if dconf read /org/gnome/desktop/interface/color-scheme | grep -q prefer-light; then
  exec delta --light "$@"
else
  exec delta --dark "$@"
fi
```

**Comment by SeanMcLoughlin** [comment]
Unfortunately none of the workarounds listed in this thread work for my workflow of:

- SSH-ing into a remote machine and attaching a tmux session
- Dynamically changing the light-dark theme of macOS and my terminal emulator depending on the time of day (I'm using iTerm2 fwiw). 

With this workflow, delta is always stuck in dark mode on the remote machine and I have to switch to system-wide dark mode in order to read diffs. I can't change my gitconfig to force delta to use light or dark mode always because I frequently change my system theme. I also can't create a wrapper script reading my system's color scheme because the problem is on the machine I've SSH'd into. 

Setting $TERM to `tmux-256color` does not work.

If I do not attach a tmux session when SSH-ing, the colors change as expected. 

It would be ideal if color theme detection would work even inside tmux.

**Comment by tautropfli** [comment]
@SeanMcLoughlin This does not seem like a delta issue, but rather an interaction issue between all the different components :)

1. delta talks to tmux (from delta's perspective tmux is the terminal) to get the current dark/light state.
2. tmux queries the connected terminal for its background color when you attach a session.

There are two things that could go wrong here:
* delta does not query tmux because tmux pretends to be screen. That's why it's important to set `TERM=tmux-256color`.
* tmux does not answer the color query because you're using an old version of tmux. You can test if tmux answers the query by running `printf '\e]10;?\e\\' && cat -v` (for the fg color) and `printf '\e]11;?\e\\' && cat -v` (for the bg color). You should see something like `^[]10;rgb:0000/0000/0000^[\` as a result.
* tmux doesn't know that the terminal has changed its colors. This happens when you are attached to tmux while your terminal changes its theme. \
   There are two ways that the terminal can inform tmux of the change:
   * Send a `SIGWINCH` signal. iTerm2 does this, but signals do not get forwarded over SSH if I recall correctly.
   * There's a [newer protocol](https://contour-terminal.org/vt-extensions/color-palette-update-notifications/) introduced by the contour terminal that works over SSH. tmux supports this, but I don't think iTerm2 does.
     * support for this protocol has landed in tmux [fairly recently](https://github.com/tmux/tmux/pull/4353) so you will need a very up-to-date version of tmux :)

What can you do?
* Detach and re-attach the tmux session after your terminal changes its theme.
* Use a terminal such as Contour, Ghostty or Kitty that supports the new contour protocol.
* [Ask](https://gitlab.com/gnachman/iterm2) iTerm2 to support the new protocol.

**Comment by SeanMcLoughlin** [comment]
@bash thank you for the detailed explanation about how all of these tools interact with one another. I guess now I have a reason to try out Ghostty.

**Comment by tautropfli** [comment]
@EbonJaeger This issue [should be fixed](https://github.com/zellij-org/zellij/issues/3590#issuecomment-4380894063) in the latest version of Zellij (0.44.2) 🎉

**Comment by EbonJaeger** [comment]
Everything appears to be working as expected without specifying `dark = true` in the `.gitconfig` now. 🎉
